File: yujung/프로그래머스/단어변환.cpp
//단어 변환

#include<iostream>
#include <string>
#include <vector>
#include<algorithm>

using namespace std;

bool visited[50] = { false, };
int answer=50;

bool check(string a, string b)
{
	int cnt = 0;
	int l = a.size();
	for (int i = 0; i < l; i++) {
		if (a[i] != b[i]) {
			cnt++;
		}
	}
	if (cnt == 1) {
		return true;
	}
	return false;
}
void dfs(string begin, string target, vector<string> words,int step) {

	if (answer <= step) {
		return;
	}
	
	if (begin == target) {
		answer = min(answer, step);
		return ;
	}
	for (int i = 0; i < words.size(); i++) {
		if (check(begin, words[i]) == true && visited[i] == false)
		{
			
			visited[i] = true;
		//	cout << begin << " " << visited[i];
			dfs(words[i], target, words,step+1);
			visited[i] = false;
			
		
		}
	}	

	return ;
}

int solution(string begin, string target, vector<string> words) {
	dfs(begin, target, words, 0);
	if (answer == 50)
		return 0;
	return answer;
}




int main() {

	string b;
	string t;
	vector<string> w;
	int n;
	string tmp;

	cin >> b;
	cin >> t;
	cin >> n;
	for (int i = 0; i < n; i++) {
		cin >> tmp;
		w.push_back(tmp);
	}
	cout<<solution(b,t, w);
	//cout << answer ;
}

File: seoyeon/프로그래머스/알고리즘 고득점 Kit/DFSBFS/단어변환.py
ans = float("inf")

def dfs(cnt,current_word,target,words):
    global ans

    if current_word == target:
        ans = min(ans,cnt)
        return ans
    
    if cnt > len(current_word)+1:
        return
    
    for w in words:
        c = 0
        for i in range(len(w)):
            if current_word[i]!=w[i]:
                c += 1
        if c==1:
            dfs(cnt+1,w,target,words)
    
def solution(begin, target, words):
    global ans
    
    #반환할 수 없는 경우
    if target not in words:
        return 0
    
    dfs(0,begin,target,words)
    
    if ans == float("inf"):
        ans = 0
    
    return ans